arch/risc-v/src/common/riscv_internal.h

```diff
@@ -253,6 +253,19 @@ void riscv_cpu_boot(int cpu);
 int riscv_pause_handler(int irq, void *c, void *arg);
 #endif
 
+/****************************************************************************
+ * Name: riscv_mhartid
+ *
+ * Description:
+ *   Context aware way to query hart id
+ *
+ * Returned Value:
+ *   Hart id
+ *
+ ****************************************************************************/
+
+uint64_t riscv_mhartid(void);
```

Comment:
why not change riscv_mhartid to macro like this:
```
#ifndef CONFIG_ARCH_USE_S_MODE
#  define riscv_mhartid() READ_CSR(mhartid)
#endif
```
So, we can catch the wrong usage at the compiler time.

Reply:
The S-mode implementation is still WIP, but I can change the function into a macro as it will work with S-mode also.

Reply:
Does it really need to be `uint64_t`? I mean maybe we can use `uintptr_t`?

Reply:
How about extend existing up_cpu_index to support S-mode instead of a new function ?
```
/* In riscv_cpuindex.c */
int up_cpu_index(void)
{
  int mhartid;

  asm volatile ("csrr %0, mhartid": "=r" (mhartid));
  return mhartid;
}
```

Reply:
Sure, did not notice that function.

I also suggest changing the return type to uintptr_t, as mhartid size is XLEN ?